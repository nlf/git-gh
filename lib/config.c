#include <stdio.h>
#include <stdlib.h>
#include <wordexp.h>
#include <json.h>
#include <json_object_private.h>
#include "config.h"
#include "jsonhelpers.h"

extern struct json_object* config_get_config() {
    int file_length;
    char* config;
    struct json_object* json;
    FILE* fp;
    wordexp_t config_path;

    json = calloc(1, sizeof(struct json_object));
    wordexp("~/.gitgh", &config_path, 0);
    fp = fopen(config_path.we_wordv[0], "r");
    wordfree(&config_path);
    if (fp == NULL) {
        printf("Invalid configuration found\nPlease run git ghsetup to create one\n");
        return NULL;
    }

    fseek(fp, 0L, SEEK_END);
    file_length = ftell(fp);
    rewind(fp);

    config = calloc(file_length + 1, sizeof(char));
    if (config == NULL) {
        fprintf(stderr, "insufficient memory\n");
        return NULL;
    }

    fread(config, file_length, 1, fp);
    fclose(fp);
    json = json_tokener_parse(config);
    free(config);
    return json;
}

extern char* config_get_token() {
    struct json_object* config = config_get_config();
    char* token = jsonh_get_string(config, "token");

    free(config);
    return token;
}

extern char* config_get_user() {
    struct json_object* config = config_get_config();
    char* user = jsonh_get_string(config, "user");

    free(config);
    return user;
}

extern int config_write_config(struct json_object* config) {
    wordexp_t config_path;
    FILE* fp;

    wordexp("~/.gitgh", &config_path, 0);
    fp = fopen(config_path.we_wordv[0], "w");
    wordfree(&config_path);
    if (fp == NULL)
        return 0;
    fputs(json_object_get_string(config), fp);
    fclose(fp);
    return 1;
}
