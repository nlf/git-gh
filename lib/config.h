extern struct json_object* readConfig();
