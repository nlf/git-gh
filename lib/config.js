var fs = require('fs'),
    path = require('path');

function parseConfig(file) {
    var parsed;
    try {
        parsed = JSON.parse(file);
    } catch (e) {
        parsed = false;
    }
    return parsed;
}

function error() {
    console.error('No valid configuration was found');
    console.error();
    console.error('To create one, please run the command: git ghsetup');
    console.error('and follow the prompts');
    process.exit(1);
}

var filePath = path.join(process.env[process.platform === 'win32' ? 'USERPROFILE' : 'HOME'], './.gitgh'),
    exists = fs.existsSync(filePath);

if (!exists) return error();

var stat = fs.statSync(filePath);
if (!stat.isFile()) return error();

var file = fs.readFileSync(filePath, 'utf8');
var config = parseConfig(file);
if (!config || !config.token || !config.user) return error();

module.exports = config;
