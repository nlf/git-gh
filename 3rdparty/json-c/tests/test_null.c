/*
* Tests if binary strings are supported.
*/

#include <stdio.h>
#include <string.h>
#include "config.h"

#include "json_inttypes.h"
#include "json_object.h"

int main() {
    // this test has a space after the null character. check that it's still included
    const char *input = " \0 ";
    const char *expected = "\" \\u0000 \"";
    struct json_object *string = json_object_new_string_len(input, 3);
    const char *json = json_object_to_json_string(string);

    int strings_match =  !strcmp( expected, json);
    int retval = 0;
    if (strings_match) {
        printf("JSON write result is correct: %s\n", json);
        printf("PASS\n");
    } else {
        printf("JSON write result doesn't match expected string\n");
        printf("expected string: ");
        printf("%s\n", expected);
        printf("parsed string:   ");
        printf("%s\n", json);
        printf("FAIL\n");
        retval=1;
    }
    json_object_put(string);
    return retval;
}
