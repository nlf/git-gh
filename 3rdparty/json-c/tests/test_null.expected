JSON write result is correct: " \u0000 "
PASS
